zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+        #[frame_support::transactional]
+        #[pallet::weight(T::WeightInfo::add_vote_outcome(T::MaxOwners::get()))]
+        pub fn add_vote_outcome(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+            outcome: OutcomeReport,
+        ) -> DispatchResultWithPostInfo {
+            let sender = ensure_signed(origin)?;
+
+            let winner_info =
+                <Winners<T>>::get(market_id).ok_or(Error::<T>::NoGlobalDisputeStarted)?;
+            ensure!(!winner_info.is_finished, Error::<T>::GlobalDisputeAlreadyFinished);
+
+            ensure!(
+                <Outcomes<T>>::get(market_id, &outcome).is_none(),
+                Error::<T>::OutcomeAlreadyExists
+            );
+
+            let voting_outcome_fee = T::VotingOutcomeFee::get();
+
+            let reward_account = Self::reward_account(&market_id);
+
+            T::Currency::transfer(
+                &sender,
+                &reward_account,
+                voting_outcome_fee,
+                ExistenceRequirement::AllowDeath,
+            )?;
+
+            Self::push_voting_outcome(&market_id, outcome.clone(), &sender, voting_outcome_fee);
+
+            Self::deposit_event(Event::AddedVotingOutcome(market_id, outcome));
+            // charge weight for successfully have no owners in Winners and no owners in empty Outcomes
+            Ok((Some(T::WeightInfo::add_vote_outcome(0u32))).into())
+        }
+
+        #[frame_support::transactional]
+        #[pallet::weight(T::WeightInfo::reward_outcome_owner(
+            T::MaxOwners::get(),
+            T::RemoveKeysLimit::get()
+        ))]
+        pub fn reward_outcome_owner(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?;
+
+            let (owners_len, removed_keys_amount) = <Winners<T>>::try_mutate(
+                market_id,
+                |winner_info: &mut Option<WinnerInfoOf<T>>| -> Result<(u32, u32), DispatchError> {
+                    let mut winner_info =
+                        winner_info.as_mut().ok_or(Error::<T>::NoGlobalDisputeStarted)?;
+                    ensure!(winner_info.is_finished, Error::<T>::UnfinishedGlobalDispute);
+
+                    // Outcome can be None when the second call happens in case RemoveKeysLimit is not reached
+                    if let Some(outcome_info) = <Outcomes<T>>::get(market_id, &winner_info.outcome)
+                    {
+                        // storage write is needed here in case, that the first call of reward_outcome_owner doesn't reward the owners
+                        // this can happen if there are more than RemoveKeysLimit keys (for KillStorageResult::SomeRemaining)
+                        winner_info.owners = outcome_info.owners;
+                    }
+
+                    let (owners_len, removed_keys_amount) = match <Outcomes<T>>::remove_prefix(
```

Comment:
`remove_prefix` is deprecated, use https://paritytech.github.io/substrate/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.clear_prefix